libs/nio/src/main/java/org/elasticsearch/nio/InboundChannelBuffer.java

```diff
@@ -232,15 +274,34 @@ private int indexInPage(long index) {
 
         private final ByteBuffer byteBuffer;
         private final Runnable closeable;
+        private final AtomicInteger referenceCount;
 
         public Page(ByteBuffer byteBuffer, Runnable closeable) {
+            this(byteBuffer, closeable, new AtomicInteger(1));
+        }
+
+        private Page(ByteBuffer byteBuffer, Runnable closeable, AtomicInteger referenceCount) {
             this.byteBuffer = byteBuffer;
             this.closeable = closeable;
+            this.referenceCount = referenceCount;
+        }
+
+        private Page duplicate() {
+            referenceCount.incrementAndGet();
```

Comment:
I am a little concerned about this. What if the referenceCount is already 0? I think this should implement `AbstractRefCounted`? I guess that is not available in this lib? I would love to move that class to `core` and replace the `AlreadyClosedException` with `IllegalStateException` and use it here?!

Reply:
I do wonder why we need to refCount this. I think it would be good to have some explanation why this is is needed. I also assume if we miss returning a page we will fail tests because not all byte blocks are released right?

Reply:
> I do wonder why we need to refCount this.

It's to integrate with netty. Netty can hold onto bytes internally and we don't know when it is down with them without overriding its buffer.

>  I also assume if we miss returning a page we will fail tests because not all byte blocks are released right?

Yeah normal mock page recycler stuff

> I think this should implement AbstractRefCounted?

I will do this but it will cause the introduction of more layers of objects. Can `Page` be reference counted? Yes. But when a return a `Page`, it needs to be a different `Page` object because the `ByteBuffer` needs to be different. So internally we will need to create like a `Releaser` object that is reference counted that is passed around to the different pages.